dbms/src/DataStreams/RuntimeFilter.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/RuntimeFilter.h>
+#include <Interpreters/Set.h>
+#include <Storages/DeltaMerge/FilterParser/FilterParser.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int SET_SIZE_LIMIT_EXCEEDED;
+}
+
+std::string RuntimeFilter::getSourceColumnName() const
+{
+    return source_column_name;
+}
+
+tipb::RuntimeFilterType RuntimeFilter::getRFType() const
+{
+    return rf_type;
+}
+
+tipb::Expr RuntimeFilter::getSourceExpr() const
+{
+    return source_expr;
+}
+
+std::string RuntimeFilter::getFailedReason() const
+{
+    return failed_reason;
+}
+
+void RuntimeFilter::setSourceColumnName(std::string & source_column_name_)
+{
+    source_column_name = source_column_name_;
+}
+
+void RuntimeFilter::setINValuesSet(const std::shared_ptr<Set> & in_values_set_)
+{
+    in_values_set = in_values_set_;
+}
+
+// There is a possibility that the same rf is canceled multiple times,
+// but rf does not have thread safety issues.
+void RuntimeFilter::updateValues(const ColumnWithTypeAndName & values, const LoggerPtr & log)
+{
+    if (isFailed())
+    {
+        return;
+    }
+    Block data(std::initializer_list<ColumnWithTypeAndName>({values}));
+    switch (rf_type)
+    {
+    case tipb::IN:
+        try
+        {
+            LOG_DEBUG(log, "update values in rf, data size:{}", data.rows());
+            in_values_set->insertFromBlock(data, true);
```

Comment:
The function ```insertFromBlock ``` is thread safe. Different join threads do update the set concurrently.

Reply:
OK, but this will make join thread blocked by each other, especially when fine grained shuffle is enabled, I think it's better to build the set concurrently?

Reply:
And why call `insertFromBlock` with `fill_set_elements` to true?

Reply:
Because I need to take the value in this in to construct the ```rs operator``` of the storage layer. If only date is stored, there is no way to get value from set.

Reply:
> OK, but this will make join thread blocked by each other, especially when fine grained shuffle is enabled, I think it's better to build the set concurrently?

When different threads update the set at the same time, we really need to use locks or segment locks to ensure the thread safety of the set. Just like different threads update the hash table.

Reply:
> Because I need to take the value in this in to construct the `rs operator` of the storage layer. If only date is stored, there is no way to get value from set.

But `set_elements` in `Set` contains all the element, not just the unique element, is that what you want?

Reply:
The unique elements should be keep in ```Set```. I will add a field to keep this values later.